src/Strategy/Dart/PubSpec.hs

```diff
@@ -0,0 +1,178 @@
+import Effect.ReadFS (Has, ReadFS, readContentsYaml)
+import Graphing (Graphing, direct, empty)
+import Path
+import Strategy.Dart.PubSpecLock (PackageName (..))
+import Types (GraphBreadth (..))
+
+data PubSpecContent = PubSpecContent
+  { pubSpecDependencies :: Maybe (Map PackageName PubSpecDepSource)
+  , pubSpecDevDependencies :: Maybe (Map PackageName PubSpecDepSource)
+  , pubSpecDependenciesOverrides :: Maybe (Map PackageName PubSpecDepSource)
+  }
+  deriving (Show, Eq, Ord)
+
+data PubSpecDepSource
+  = PubSpecDepHostedSource
+      { version :: Maybe VerConstraint
+      , hostedName :: Maybe Text
+      , hostedUrl :: Maybe Text
+      }
+  | PubSpecDepGitSource
+      { gitRef :: Maybe VerConstraint
+      , gitUrl :: Text
+      }
+  | PubSpecDepSdkSource {sdkName :: Text}
+  | PubSpecDepPathSource {hostPath :: Text}
+  deriving (Show, Eq, Ord)
+
+instance FromJSON PubSpecContent where
+  parseJSON = Yaml.withObject "pubspec.yaml content" $ \o -> do
+    dependencies <- o .:? "dependencies"
+    devDependencies <- o .:? "dev_dependencies"
+    depOverrides <- o .:? "dependency_overrides"
+    pure $ PubSpecContent dependencies devDependencies depOverrides
+
+-- Formatter crashes here.
+{- ORMOLU_DISABLE -}
+instance FromJSON PubSpecDepSource where
+  parseJSON (Yaml.String s) = pure $ PubSpecDepHostedSource (Just . CEq $ s) Nothing Nothing
+  parseJSON (Yaml.Object o) =
+    asum
+      [ PubSpecDepHostedSource . Just . CEq
+          <$> o .: "version"
+          <*> o .: "hosted" |> "name"
+          <*> o .: "hosted" |> "url"
+      , PubSpecDepGitSource . Just . CEq
+          <$> o .: "git" |> "ref"
+          <*> o .: "git" |> "url"
+      , PubSpecDepGitSource Nothing
+          <$> o .: "git"
+      , PubSpecDepPathSource
+          <$> o .: "path"
+      , PubSpecDepSdkSource
+          <$> o .: "sdk"
+      ]
+    where
+      (|>) :: FromJSON a => Yaml.Parser Yaml.Object -> Text -> Yaml.Parser a
+      (|>) parser key = do
+        obj <- parser
+        obj .: key
+  parseJSON _ = fail "failed parsing pub package's source!"
+{- ORMOLU_ENABLE -}
+
+toDependency :: [DepEnvironment] -> PackageName -> PubSpecDepSource -> Dependency
+toDependency environments name (PubSpecDepHostedSource version _ url) =
+  Dependency
+    { dependencyType = PubType
+    , dependencyName = unPackageName name
+    , dependencyVersion = version
+    , dependencyLocations = maybeToList url
+    , dependencyEnvironments = environments
+    , dependencyTags = Map.empty
+    }
+toDependency environments _ (PubSpecDepGitSource gitRef gitUrl) =
+  Dependency
+    { dependencyType = GitType
+    , dependencyName = gitUrl
+    , dependencyVersion = gitRef
+    , dependencyLocations = []
+    , dependencyEnvironments = environments
+    , dependencyTags = Map.empty
+    }
+toDependency _ _ (PubSpecDepSdkSource _) = error "unexpected sdk dependency conversion"
+toDependency _ _ (PubSpecDepPathSource _) = error "unexpected path dependency conversion"
```

Comment:
We can't use `error` here -- it'll just throw a runtime exception. Instead, we could change the type of `toDependency` to express that it can fail, e.g. with `Maybe`:

```hs
toDependency :: [DepEnvironment] -> PackageName -> PubSpecDepSource -> Maybe Dependency
```

This has ramifications for its callers; specifically from my other comment:

```hs
-- this won't work anymore; has the wrong type
finalGraph :: Graphing Dependency
finalGraph = Graphing.gmap toDependency ...
```

luckily there is a workaround:
```hs
-- the real type now that `toDependency`'s type has changed
finalGraph1 :: Graphing (Maybe Dependency)
finalGraph1 = Graphing.gmap toDependency ...

-- we can use the `induceJust` function to filter out the `Nothing`s
finalGraph2 :: Graphing Dependency
finalGraph2 = Graphing.induceJust finalGraph1
```